### ISSUE TITLE
millions of items deleted using ME io port

### ISSUE THREAD

**Issue by T0MSSM0T** [AppliedEnergistics/Applied-Energistics-2#4013]
TLDR: possible broken cables but still visually showing to be connected (or maybe not since I didn't examine every cable before replacing them) caused an ME IO Port to move items from my drives into nothing, instead of my barrels, causing them to be deleted, after replacing cables and ME IO Port it worked as intended (not sure what fixed it out of those 2 most likely the cables though)

**Describe the bug**

To explain my ME system on the necessary parts:
MANY 64k drives (around 50 empty  64k drives) set to prio 80, 70, 60, 50
Storage bus connected to a Item barrel connector set to prio 1000 (the barrels of the items in question were NOT locked)
Lots of auto crafting (interfaces, molecular assemblers, endio machines, mekanism maachines ect )
and that is about it, no trash can, no export buses, no other place items should be able to go
especially not millions of items (I haven't checked every machine but it should be impossible for all my items to fit in the around 20 machines).

I used the ME IO port to transfer my items from drives into my Yabba barrels.
I saw the number decreasing while hovering over my drives and then then getting moved into the "done" slot, while my drives got filled up with non barrel items and i thought my barrels were filling up as well.
I didn't check it right away, eventually I looked saw nothing was there and then tried to move the items again (there were deleted by now but I didn't realize) using the ME IO Port.

**To Reproduce**

Since I am not exactly sure of the issue its hard to reproduce.
If I had to guess cables some how broke but didn't break.
What I mean is every once in a while when I go into the simple void dimension I hear glass breaking (I have glass and cables that can be the source) but one time I heard it I looked around and saw "glitched out" cable it was not broken but missing some of the connections even though it should connect but the cable still worked, giving all the connected ME things power and channels.
This is the only thing I can think of that could have happens, it saw the Storage bus but it wasn't really connected, items that already had a place were still going into the barrels no problem though.
I am sorry I can't exactly help with the problem, and I know the amount of items and what items I lost so I can fix my problem, but this just seems like a huge issue, I lost about 5 million items.

I have replaced the cable and everything and the ME IO Port works now.
I also moved the ME IO Port and it still works, not sure which fixed or if both were required.

Before I realized my items didn't get move/were deleted I started running into a problem were my ME system was having weird problems, import buses weren't ALWAYS working, like they would work, then freeze, and repeat, also I started testing export bus craft card and redstone card and level emitter and I would get 2s of lag while it was on (opening up a chest or inv would take about 2s) so I replaced these cables as well and it worked.

If anyone has anything they want to know or anything I can try or look at let me know.

**Expected behavior**

The ME IO Port move my items from the drives into the barrels OR not move anything at all.

**Additional context**

The Real problem might just be cables need to not be broken but still be connected, it is really hard to say what to do, the main thing I am trying to accomplish with this report is let an issue be known.

**Environment**

I am playing Project Ozone 3 mod pack 

- Minecraft Version: 1.12.2
- Ozone version: 3.0.40
- AE2 Version: appliedenergistics2-rv6-stable-6.jar
- Forge Version: 14.23.5.2836

**Comment by mindforger** [comment]
>Storage bus connected to a Item barrel connector set to prio 1000 (the barrels of the items in question were NOT locked)

did you cross check if items still vanish when you just disconnect them, leaving only the drives attached ... just to rule out it's nothing similar to the old "locked DSU" bug that may popped up

since you said it is hard to reproduce without having your setup it would be immensely helpfull if you can modify your system just bit by bit until you notice the items vanishing again to narrow down the issue? The modpack itself is way to big to put it into debug

**Comment by yueh** [comment]
IO Ports won't remove items before they are accepted by another inventory. If they void them, it is usually because the player attached a storage bus to a trashcan and forget about it or there are a couple inventories around, which will destroy them immediately. Like @mindforger mentioned locked DSUs will act as trashcan.

**Comment by T0MSSM0T** [comment]
I haven't built a trash can, i have saved every single piece of armor from a mob farm I have. also only 1 single storage bus exists.
I streamed it live and can link the VOD as well as some specific times in the VOD to look at.
They were all UNLOCKED infinite upgraded barrels.

The problem is before I realized they were deleted I thought they were still in my drives so I decided to break and replace all my cables so its hard to reproduce.

But the VOD can maybe help figure the problem out.

**Comment by T0MSSM0T** [comment]
@mindforger 
@yueh

**Comment by Ommina** [comment]
Have you also left an issue on the YABBA issue page?  If not, how have you determined this represents a problem with the sending tile (IOBus) vs the receiving tile (Connector)?

I can understand the oddness you were experiencing with the lag leading you here, but given the two way nature of item transfers from one tile to another, I would be unsurprised if the actual cause was unrelated to either mod.

Which, I admit, isn't much of a consolation.

**Comment by mindforger** [comment]
the video may not show anything helpfull, like i said it would be up to you to check if the issue persists when you start unplugging parts of the storage area to narrow down the source of loss

**Comment by T0MSSM0T** [comment]
@Ommina you have a point, but as i mentioned I personally dont care that much, I am just cheating all my items back in.
The main reason I posted this issue and why I posted it here is because I believe it is very likely AE2.
And although I haven't tested it yet (maybe I can try to bug it right now) I believe it is because the cable glitch broke but didn't actually break.

what everyone seems to be missing is how simple my AE2 set up is. and how I looked into everything and the items just have had to gone missing. it was only the 17 items I choose to... not every item from the drives is missing. just the 17 items I wanted to move into barrels.

I am merely here responding to try to help find the problem, I am not pointing fingers at AE2 or anyone, but if this happened for real on someone who didn't record they would probably quit.

@mindforger although maybe more work then you would like would have to be done, but if you watched enough you might be able to see my whole system and maybe you can point out where my items went, or maybe you can point out an error in something I did.
Also I am not sure I follow you PS statement.

**Comment by T0MSSM0T** [comment]
Here, I just got it to happen, the cable I am moused over wasn't there when it "broke" but as you can see... the import bus still is online even though it visually is broken. I have also tried the specific one that looks unconnected and it does import, but I wonder if this happens enough, what might actually happen? (trying to cause multiple problems currently)

After reloading the world the cable comes back.

Once again I will say I am not sure of what the problem is nor do I blame AE2, but there is no place my 17 items could have gone, no trash can (or anything similar) and I have checked all the chests just in case and nothing is in any of them, I have also taken an item that went missing (tomato) and put it into each machine I had to see if they could be destroyed and nope.

![2019-04-23_02 05 31](https://user-images.githubusercontent.com/15080064/56558412-a45de100-656c-11e9-843f-e83ef77d3e4b.png)

**Comment by mindforger** [comment]
>but if you watched enough

i would do if there was a link to the video ... or did i just overlook the link anywhere ?! (i just went through the topic two times again to make sure but i can be blind sometimes)

**Comment by T0MSSM0T** [comment]
didn't want to post the link until someone asked, don't want to self advertise
https://www.twitch.tv/videos/414453629

around 1:08:30 I start the putting 17 items into barrels
around 2:26:00 i start getting bus lags, and 2:28:00 I start to look into what is going on

**Comment by mindforger** [comment]
>didn't want to post the link until someone asked, don't want to self advertise

sure i understand that :) is added some text to my last post, just to let you know :P

**Comment by T0MSSM0T** [comment]
I can also provide a world download, nothing much as changed since this issue has came up.
I power it with Draconic Evolution wireless power, 260k rf per tick transfer I believe.
But finding the source of this is my goal, regardless if it is actually user error, or yabba or ae or something else entirely.

**Comment by T0MSSM0T** [comment]
feel free to ask questions about the video which might not be clear.

**Comment by mindforger** [comment]
i have to delay the watching of the video as i currently only have very limited internet, maybe later that evening when i get to borrow some wifi

**Comment by T0MSSM0T** [comment]
sure thing!

**Comment by yueh** [comment]
The main point here is that it only happened to items moved to the barrel. And all we can do is to rely on what some external inventory tells us. If they say they stored it, but didn't, it is beyond our control.

I skimmed through the yabba code and there are a couple of questionable things. Like it probably will kill your world once too many barrels are linked with a `StackOverflowError` including each time you want to load it afterwards. Handling chunk loading/unloading seems to be a bit questionable. E.g. if I did not overlook something, in case a barrel is unloaded, they controller thingy still refers to it as loaded and probably will still accept items for them. Which will then never be saved and just overwritten once the chunk is reloaded. If that even reliably happens and in does not forever refer to the old one until replaced in some rare cases.

**Comment by T0MSSM0T** [comment]
My barrels and my me io port are within 30 blocks of each other.

Also is that how the AE code works? "If they say they stored it, but didn't," if that is the case then yes it isn't AE2

Also i had about 2k barrels placed before and it didn't cause a problem except lag, now i have about 100 so I don't think that is the problem.

**Comment by T0MSSM0T** [comment]
Note, I just moved about 1.5 million items in the same way and it worked, the only difference is replaced the cable, and I guess hidden things that are not apparent like chunk loading (although I very highly doubt it is that)

**Comment by mindforger** [comment]
watching the video, i got a few things to add:

despite having a lot of controllers (which increase the system capacity slightly) you should add a few dense energy cells, because the input is limited by capacity/tick and you have A LOT item movement which puts your me system in bootloop (and causing the lagging)

also the chicken chunks chunkloaders are a bit edgy, try dropping a spotloader on top of that chunkloader, they work more reliably in this configuration

you may had a slightly glitched me system through a heavy lagged game probably

**Comment by T0MSSM0T** [comment]
I will do that, but is lag shouldn't cause the problems I had, I don't think.

I will look at the wiki what do energy cells do that my DE crystal wont?

**Comment by mindforger** [comment]
they provide internal capacity to your me system ... if your system runs out of internal juice in a single tick, it shuts down and reboots

this limits the input

**Comment by T0MSSM0T** [comment]
alright, sounds like a smart idea. anything else you can think of or is the most likely culprit the barrels said they took the items but didnt?

**Comment by mindforger** [comment]
i don't see anything in the video that screams for issues so far, still scrolling through the video, but the amount of item movement could indicate power issues ... also the addons like extracells do things to the me system, one reason i never use them and have no idea what effects they may cause

**Comment by T0MSSM0T** [comment]
what from extra cells did i use?

**Comment by mindforger** [comment]
the general presence of mods that interact with ME can change behaviour, you actually do not need to use things from it (i am very conservative in using mods, a reason why i don't like addon mods that much, it's just a wrench on a rusty hook ready to fall into the gears :)

**Comment by T0MSSM0T** [comment]
I share a similar opinion! but mod packs I don't want to change them one reason why I only use 64k drives enough though i can make the  16XXXk drives
